### PR TITLE
Improve Windows compatibility

### PR DESCRIPTION
**Pull request by TheBrenny**
Added a method to acquire the session history on windows. It doesn't get the history, and also doesn't check to see if the user has a HISTFILE environment variable. But I don't know if any windows user actually has one for CMD.

This won't work if invoked through PowerShell though, because PS uses the history command.

🙏

### CODE REVIEW

**Comment by sindresorhus** [comment]
I don't really understand the point. It gets only the session history, but not permanent history? However, the session history is nothing as we're spawning the terminal. Can you please elaborate?

**Comment by TheBrenny** [comment]
I know it doesn't get a persistent history, but getting the session's history is also a common use case (I imagine). Otherwise, if a HISTFILE env variable is set, that file should be used - because there are tools out there which persist CMD history across sessions.

Otherwise, a session history use case would be:
```console
$ git clone https://github.com/sindresorhus/shell-history.git && alert
```
Where `alert` is an alias to `node C:\\alert.js` (for example), `alert.js` can read the session's history to make a toast notification with "git completed!" as the title.

I hope that makes more sense.

**Comment by sindresorhus** [comment]
> but getting the session's history is also a common use case (I imagine).

But does it actually get the history of the whole session? Like commands executed before `shell-history` was running.

File: index.js [hunk 1]
```diff
@@ -51,7 +51,9 @@ module.exports = opts => {
 	opts = opts || {};
 
 	if (process.platform === 'win32') {
-		return [];
+		let cp = require("child_process");
+		let dk = cp.spawnSync("doskey", ["/history"]);
```
Comment (sindresorhus) [inline index.js:55]: Use a more descriptive name than `dk`.

Comment (sindresorhus) [reply index.js:55]: Use the `encoding` option instead of `toString()`.

File: index.js [hunk 2]
```diff
@@ -51,7 +51,9 @@ module.exports = opts => {
...
+		return dk.stdout.toString().replace(/\n/g, '').split('\r') || [];
```
Comment (sindresorhus) [inline index.js:56]:
```suggestion
		return dk.stdout.trim().replace(/\n/g, '').split('\r') || [];
```

Comment (sindresorhus) [reply index.js:56]: Why are you removing `\n`? Does it output `\n\r` or `\r` if so, it's better to just split on either.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Please always run `npm test` locally before committing.

**Comment by sindresorhus** [comment]
The special Windows behavior (of only getting the session history) needs to be documented.

File: index.js (hunk 2 above)
Comment (TheBrenny) [reply index.js:56]: Noted and changed.

File: index.js (hunk 1 above)
Comment (TheBrenny) [reply index.js:55]: Noted and changed.

`toString()` is required because stdout is a buffer.

**Comment by TheBrenny** [comment]
> Please always run `npm test` locally before committing.

I tried running the tests but it just isn't working. First, there are issues because `eslint-plugin-ava` causes an `undefined` TypeError to throw, because [that's how it's made](https://github.com/avajs/eslint-plugin-ava/blob/master/rules/no-identical-title.js#L32-L35). I did a clean install on WSL (which defeats the purpose of the test) which failed anyway because of the previous issue.

I then decided to skip the whole `xo` process (because none of Node12(Win), Node12(Linux) or Node14(Linux)) worked.

Calling `node_modules/.bin/ava` (both the `sh` and `cmd` variants on Win and WSL) didn't work because test.js uses ES6 syntax, which I'm not compatible with apparently...

At the end of the day, I converted test.js to Node's `require` syntax, called `node_modules/.bin/ava` and magic almost happened - I gave the test a name, and *then* it worked with 1 test passed.

I didn't plan on giving you my life story on trying to run XO and Ava, so sorry about that! 😅

> The special Windows behavior (of only getting the session history) needs to be documented.

Doneskies.

Send feedback! 🎉

**Comment by sindresorhus** [comment]
> I then decided to skip the whole xo process (because none of Node12(Win), Node12(Linux) or Node14(Linux)) worked.

XO works fine on all those setups. This is a problem with your setup. But you can look at the lint failures in the Travis output and fix manually.

File: index.js [hunk 3]
```diff
@@ -19,39 +19,45 @@ function parse(str) {
...
+		paths.add(path.join(homeDir, '.zsh_history'));
+		paths.add(path.join(homeDir, '.history'));
+	}
+
 	return Array.from(paths)
 		.filter(fs.existsSync)
 		.map(fp => ({fp, size: fs.statSync(fp).size}))
-		.reduce((a, b) => a.size > b.size ? a : b).fp;
+		.reduce((a, b) => a.size > b.size ? a : b, {
+			fp: null
```
Comment (sindresorhus) [inline index.js:47]: Not clear why this change is required.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/shell-history/pull/5#discussion_r502087899 is not resolved. Please read my comment again. I think it's quite clear.

File: index.js (hunk 3 above)
Comment (TheBrenny) [reply index.js:47]: `reduce` was throwing an error because of how Windows is being handled - Win platforms don't resolve early, instead, there are no paths to `reduce`, thus causing an error, or requiring a default value.

IMO, after filtering and mapping, the first path should be used, which can allow the use of:
```js
Array.from(paths).filter(fs.existsSync)[0]
```
instead of mapping and reducing as well. If index 0 doesn't exist, it returns undefined which is handled [here](https://github.com/sindresorhus/shell-history/pull/5/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R56).

**Comment by TheBrenny** [comment]
> > but getting the session's history is also a common use case (I imagine).
> 
> But does it actually get the history of the whole session? Like commands executed before `shell-history` was running.

Sorry I missed this comment -- I've added `test_win.bat` which shows how this works. I don't know how to turn that into a "test" per se, but it's a visual tester I guess... The only bug which I know of for this is when using `clink` (which overrides doskey, so it's a problem on that end, not here).

<hr>

#6 Fixed my problems (big ty @RocktimSaikia 🎉) -- Test now passes correctly.

<hr>

> [#5 (comment)](https://github.com/sindresorhus/shell-history/pull/5#discussion_r502087899) is not resolved. Please read my comment again. I think it's quite clear.

You are correct it wasn't resolved - I didn't realise that `spawnSync` had an encoding option. (Off-topic, but what's the benefit over `toString` if its default behaviour is to encode to utf8?)

This has now been resolved.

File: readme.md [hunk 4]
```diff
@@ -22,19 +22,22 @@ console.log(shellHistory.path());
...
 Get an array of commands.
 
-On Windows it will always be an empty array as command history is not persisted there.
+On Windows, unless the `HISTFILE` environment variable is set, this will only return commands from the current session.
 
 ### shellHistory.path()
 
 Get the path of the file containing the shell history.
 
+On Windows, this will return either the `HISTFILE` environment variable, or null.
```
Comment (sindresorhus) [inline readme.md:39]: I don't see any `null` in the code.

File: index.js [hunk 5]
```diff
@@ -63,7 +60,17 @@ module.exports = options => {
 	options = options || {};
 
 	if (process.platform === 'win32') {
-		return [];
+		const histFile = getPath(options);
```
Comment (sindresorhus) [inline index.js:63]: The naming is an implementation detail of the `getPath` method.
```suggestion
		const historyPath = getPath(options);
```

File: index.js [hunk 6]
```diff
@@ -63,7 +60,17 @@ module.exports = options => {
...
+		if (histFile) {
+			return parse(fs.readFileSync(histFile, 'utf8'));
+		}
+
+		// No histfile used -- spawn a doskey
+		const childProcess = require('child_process');
+		const doskey = childProcess.spawnSync('doskey', ['/history'], {
```
Comment (sindresorhus) [inline index.js:70]:
```suggestion
		const {stdout} = childProcess.spawnSync('doskey', ['/history'], {
```

File: index.js [hunk 7]
```diff
@@ -63,7 +60,17 @@ module.exports = options => {
...
+			encoding: 'utf8'
+		});
+		return doskey.stdout.trim().split('\r\n') || []; // Enforce array
```
Comment (sindresorhus) [inline index.js:73]: The comment is moot. The code is clear enough.

File: index.js [hunk 8]
```diff
@@ -63,7 +60,17 @@ module.exports = options => {
...
 
 	if (process.platform === 'win32') {
-		return [];
+		const histFile = getPath(options);
+		if (histFile) {
+			return parse(fs.readFileSync(histFile, 'utf8'));
+		}
+
+		// No histfile used -- spawn a doskey
+		const childProcess = require('child_process');
```
Comment (sindresorhus) [inline index.js:69]: Imports should be at the top of the file.

File: test_win.bat [hunk 9]
```diff
@@ -0,0 +1,12 @@
+:: Acts as a windows tester.
+
+@echo off
+
+echo Print entire history
+node -p "const history = require('./'); history()"
```
Comment (sindresorhus) [inline test_win.bat:6]:
```suggestion
node -p "const history = require('.'); history()"
```

File: test_win.bat [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+pause
+echo.
+echo Print last command (which will be "test_win.bat")
+node -p "const history = require('./'); let h = history(); h[h.length-1]"
```
Comment (sindresorhus) [inline test_win.bat:10]:
```suggestion
node -p "const history = require('.'); let h = history(); h[h.length - 1]"
```

File: test_win.bat [hunk 11]
```diff
@@ -0,0 +1,12 @@
+:: Acts as a windows tester.
```
Comment (sindresorhus) [inline test_win.bat:1]: This sentence is not clear enough and it's `Windows`.

File: index.js [hunk 12]
```diff
@@ -19,10 +19,7 @@ function parse(string) {
 function getPath(options) {
 	options = options || {};
 
-	if (process.platform === 'win32') {
-		return '';
-	}
-
+	// HISTFILE might be cross-platform, so win32 check is after!
```
Comment (sindresorhus) [inline index.js:22]: "might" is not a good explanation. Explain exactly when it would be.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
> I don't know how to turn that into a "test" per se, but it's a visual tester I guess...

Then that should be explained.

**Comment by sindresorhus** [comment]
> Off-topic, but what's the benefit over toString if its default behaviour is to encode to utf8?)

No technical benefit in this case. But it's more readable to declare it upfront. It does matter when it comes to streams though, as it might end up splitting Unicode characters incorrectly.

**Comment by sindresorhus** [comment]
Please spend some more time perfecting this pull request.

File: .gitignore [hunk 13]
```diff
@@ -1,2 +1,3 @@
 node_modules
-yarn.lock
\ No newline at end of file
+yarn.lock
+package-lock.json
```
Comment (sindresorhus) [inline .gitignore:3]: Don't do unrelated changes.

File: readme.md [hunk 14]
```diff
@@ -22,19 +22,20 @@ console.log(shellHistory.path());
 //=> '/Users/sindresorhus/.history'
 ```
 
-
```
Comment (sindresorhus) [inline readme.md:25]: Don't do unrelated changes.

File: readme.md [hunk 15]
```diff
@@ -22,19 +22,20 @@ console.log(shellHistory.path());
...
 Get an array of commands.
 
-On Windows it will always be an empty array as command history is not persisted there.
+On Windows, unless the `HISTFILE` environment variable is set, this will only return commands from the current session.
 
 ### shellHistory.path()
 
 Get the path of the file containing the shell history.
 
+On Windows, this will return either the `HISTFILE` environment variable, or `undefined`.
```
Comment (sindresorhus) [inline readme.md:37]:
```suggestion
On Windows, this will return either the `HISTFILE` environment variable or `undefined`.
```

File: index.js [hunk 16]
```diff
@@ -63,7 +60,16 @@ module.exports = options => {
...
+		const historyPath = getPath(options);
+		if (historyPath) {
+			return parse(fs.readFileSync(historyPath, 'utf8'));
+		}
+
+		// No histfile found, so spawn a doskey
+		const {stdout} = childProcess.spawnSync('doskey', ['/history'], {
+			encoding: 'utf8'
+		});
+		return stdout.trim().split('\r\n') || [];
```
Comment (sindresorhus) [inline index.js:69]:
```suggestion
		return stdout.trim().split('\r\n');
```

The last part will never be hit.

File: test_win.bat [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+@echo off
+
+echo Print entire history
+node -p "const history = require('.'); history()"
+pause
+echo.
+echo Print last command (which will be "test_win.bat")
+node -p "const history = require('.'); let h = history(); h[h.length - 1]"
+pause
+
```
Comment (sindresorhus) [inline test_win.bat:14]:
```suggestion
```

File: test_win.bat [hunk 18]
```diff
@@ -0,0 +1,14 @@
+:: Confirms that the batch file works correctly in Windows Terminals/Command Prompts
+:: This batch file is used to visually test to see if shell-history works, rather
+:: than to act as a real, responsive "test case" (as such, ava is not used here)
```
Comment (sindresorhus) [inline test_win.bat:3]:
```suggestion
:: than to act as a real, responsive "test case" (as such, `ava` is not used here).
```

File: test_win.bat [hunk 19]
```diff
@@ -0,0 +1,14 @@
+:: Confirms that the batch file works correctly in Windows Terminals/Command Prompts
+:: This batch file is used to visually test to see if shell-history works, rather
```
Comment (sindresorhus) [inline test_win.bat:2]:
```suggestion
:: This batch file is used to visually test to see if `shell-history` works, rather
```

File: test_win.bat [hunk 20]
```diff
@@ -0,0 +1,14 @@
+:: Confirms that the batch file works correctly in Windows Terminals/Command Prompts
```
Comment (sindresorhus) [inline test_win.bat:1]:
```suggestion
:: Confirms that the batch file works correctly in Windows Terminal / Command Prompt.
```

File: test_win.bat [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+:: This batch file is used to visually test to see if shell-history works, rather
+:: than to act as a real, responsive "test case" (as such, ava is not used here)
+
+@echo off
+
+echo Print entire history
+node -p "const history = require('.'); history()"
```
Comment (sindresorhus) [inline test_win.bat:8]: Use the long-flag of `-p`.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Bump
